### DISCUSSION TITLE
Read data path and performance with higher replica counts.

### DISCUSSION THREAD

**Discussion by bluke** [longhorn/longhorn#4890] [Q&A] [answered]
Hi everyone.

For academic purposes, we have been running IO benchmarks on longhorn volumes, in hopes to establish a baseline for the behavior expected of this kind of in cluster distributed storage solution.  For the most part, we have been able to match our observations with the explanations provided by the doc available on longhorn.io, which is quite exhaustive.

However, we noted that bandwidth and io/s of Reads were negatively impacted by the number of replicas of a volume, and we have not been able to find a reason as to why in the documentation. The documentation has a section on reads and writes, but it concerns the volume layering and how it interacts with snapshots. We did not find any indication as to why read operations interrogate all replicas rather than a single one. Notably, we are curious whether longhorn uses this opportunity to perform additional integrity checks, or what would happen if the replicas disagreed on the content of a given block.

If someone familiar with the inner workings of longhorn has an idea, we would be happy to hear from you.
Otherwise, a pointer to what piece of code manages reads on the engine would be helpful, since we are not yet fluent enough in Golang to figure it out ourselves.

**Comment by shuo-wu** [comment, marked as answer]
Longhorn read [fetches data from all running replicas in a round-robin way](https://github.com/longhorn/longhorn-engine/blob/acdfe5be79f75acabb49a39105643752d601495f/pkg/controller/replicator.go#L115), which actually decrease the potential latency between 2 requests and results in better read performance.

As for the integrity check, or in other words, the consistency among running replicas, currently there is no extra check. Since each write operation will be confirmed only when all replicas finish the write. When there is replica failed to handle a write, the engine, which is the controller of all replicas, will mark that replica as state ERR. 
But in the next release, Longhorn will calculate and record the checksum for the snapshots in all running replicas. This is for detecting underlying storage bit rots. You can check [this ticket](https://github.com/longhorn/longhorn/issues/3198) for this part. I don't think this bit-rot detection feature is related to the replica impact against the Longhorn volume read performance.

**Reply by bluke** [reply]
Great, thank you for your help.
